examples/gazebo_iris/gazebo_quadcopter_plugin.cc

```diff
@@ -0,0 +1,391 @@
+    this->m_RollPID.SetCmdMin(param);
+    this->m_PitchPID.SetCmdMin(param);
+    this->m_YawPID.SetCmdMin(param);
+
+    this->m_RollPID.SetCmd(0);
+    this->m_PitchPID.SetCmd(0);
+    this->m_YawPID.SetCmd(0);
+    
+    char* logfile_location=std::getenv("LOG_FILE");
+    if(logfile_location!=NULL){
+        this->m_Logfile.open(logfile_location, std::ios_base::app);
+    }
+}
+
+/////////////////////////////////////////////////
+void
+GazeboQuadCopterPlugin::OnMsg(ConstQuaternionPtr &_msg)
+{
+    std::lock_guard<std::mutex> lock(this->m_Mutex);
+    this->m_Thrust = _msg->w();
+    this->m_Roll = _msg->x();
+    this->m_Pitch = _msg->y();
+    this->m_Yaw = _msg->z();
+}
+void 
+GazeboQuadCopterPlugin::MotorMixing(const double _dt)
+{
+    float thrust=0, roll=0, pitch=0, yaw=0;
+    float fr=0, fl=0, br=0, bl=0;
+    double z, r, p;
+    ignition::math::Pose3d actualPose, referencePose;
+    actualPose = this->m_Model->WorldPose();
+    referencePose = this->m_LoiterReference;
+    z = actualPose.Pos().Z();
+    r = actualPose.Rot().Roll();
+    p = actualPose.Rot().Pitch();
+
+    if(0.49 < this->m_Thrust && this->m_Thrust < 0.51 ){
+        thrust = this->m_ThrustPID.Update(z - referencePose.Pos().Z(), _dt);
+    }
+    else{
+        thrust = this->m_Thrust;
+        this->m_LoiterReference.Set(referencePose.Pos().X(), referencePose.Pos().Y(), z, referencePose.Rot().Roll(), referencePose.Rot().Pitch(), referencePose.Rot().Yaw());
+    }
+
+    roll = this->m_RollPID.Update(r - this->m_Roll, _dt); //angular displacement PID
+    pitch = this->m_PitchPID.Update(p - this->m_Pitch, _dt); // angular displacement PID
+    yaw = this->m_Yaw*0.1; //angular velocity PID
+
+    if(this->m_Logfile.is_open()){
+        this->m_Logfile << "thrust: " << thrust <<"\troll: " <<roll << "\tpitch: " <<pitch << "\tyaw: " <<yaw<< "\t"; 
+        this->m_Logfile << "\tpos: " << actualPose.Pos() <<" \trot: " <<actualPose.Rot() << std::endl; 
+    }
+    fr = thrust + yaw - pitch - roll;
+    fl = thrust - yaw - pitch + roll;
+    br = thrust - yaw + pitch - roll;
+    bl = thrust + yaw + pitch + roll;
+
+    this->m_Rotors[0].m_Cmd = this->m_Rotors[0].m_MaxRpm * fr;
+    this->m_Rotors[1].m_Cmd = this->m_Rotors[1].m_MaxRpm * bl;
+    this->m_Rotors[2].m_Cmd = this->m_Rotors[2].m_MaxRpm * fl;
+    this->m_Rotors[3].m_Cmd = this->m_Rotors[3].m_MaxRpm * br;
+}
+/////////////////////////////////////////////////
+void GazeboQuadCopterPlugin::OnUpdate()
+{
+  std::lock_guard<std::mutex> lock(this->m_Mutex);
+
+  gazebo::common::Time curTime = this->m_Model->GetWorld()->SimTime();
+
+    this->ApplyMotorForces((curTime -this->m_LastControllerUpdateTime).Double());  
+    this->m_LastControllerUpdateTime = curTime;
+}
+/////////////////////////////////////////////////
+void
+GazeboQuadCopterPlugin::ResetPIDs()
+{
+    // Reset velocity PID for rotors
+    for (size_t i = 0; i < this->m_Rotors.size(); ++i) {
+        this->m_Rotors[i].m_Cmd = 0;
+    }
```

Comment:
Another way to do this would be:
```c++
std::fill(m_Rotors.begin(), m_Rotors.end(), 0);
```

If you haven't come across the STL algorithms before, this is a good talk about them: https://www.youtube.com/watch?time_continue=1&v=2olsGf6JIkU (but don't feel obligated to watch it!)
They have a weird syntax but are really useful once you get past that.

Reply:
Ok, but won't this set all the members to 0? I mean we only need to reset the `m_Cmd` member. I don't know how `std::fill` operates on objects.

Reply:
Oops, I misread it. Yes you're totally right, it wouldn't work. My bad.